EIPS/eip-8037.md

```diff
@@ -31,62 +31,133 @@ The relationship we are seeing in this example is not linear as expected. This i
 
 Besides the parameter changes, this proposal introduces an independent metering for the code deposit costs. The specification is derived from [EIP-8011](./eip-8011.md). However, it only requires two dimensions, namely, `gas` and `code_deposit_gas`.
 
 ### Contract deployment cost calculation
 
-This proposal further changes how contract deployment cost are computed. When a contract creation returns a runtime bytecode `R` (length `L`), first check whether the code already exists in the state trie. Then:
-
-- If `CodeExists(keccak256(R), live_state) == false`, charge `code_deposit_cost=GAS_CODE_DEPOSIT*L` and store `R` under its code hash.
-- If `CodeExists(...) == true`, do not charge code-deposit storage gas (`code_deposit_gas=0`); simply link the new account's `codeHash` to the existing code object.
-
-In addition, contract creation is also charged a `storage_access_cost = GAS_WARM_ACCESS (if warm) | GAS_COLD_SLOAD (if cold)` and a `hash_cost = GAS_KECCAK256_WORD * code_data_words`, where `code_data_words = ceil(L / 32)`. This accounts for the added execution cost of accessing and verifying for duplicated code.
+This proposal fundamentally changes how contract deployment costs are computed to properly account for success and failure paths, ensuring that post-execution costs are only charged when deployment succeeds.
+
+#### Success vs Failure Gas Accounting
+
+When a contract creation transaction or opcode (`CREATE`/`CREATE2`) is executed, gas is charged differently based on whether the deployment succeeds or fails. Given runtime bytecode `R` (length `L`) returned by initcode and `H = keccak256(R)`:
+
+1. **On opcode entry:** Always charge `CREATE_FIXED_COST` (212,800 gas)
+2. **During initcode execution:** Charge `EXECUTION_GAS` equal to gas consumed by the initcode
+3. **Success path** (no error, not reverted, and `L ≤ MAX_CODE_SIZE`):
+   - If the target account is new, charge `ACCOUNT_CREATION_COST` (212,800 gas)
+   - Charge `HASH_GAS(L)` where `HASH_GAS(L) = 6 × ceil(L / 32)` to compute `H`
+   - Perform deduplication check (see Access-List CodeHash section below):
+     - If duplicate: do NOT charge `CODE_DEPOSIT_GAS(L)`; link `codeHash` to `H`
+     - If not duplicate: charge `CODE_DEPOSIT_GAS(L)` where `CODE_DEPOSIT_GAS(L) = 1,900 × L` and persist `R` under `H`, then link
+4. **Failure paths** (REVERT, OOG/invalid during initcode, OOG during code deposit, or `L > MAX_CODE_SIZE`):
+   - Do NOT charge `ACCOUNT_CREATION_COST`, `HASH_GAS(L)`, or `CODE_DEPOSIT_GAS(L)`
+   - No code is stored; no `codeHash` is linked to the account
+   - The account remains unchanged or non-existent
+
+**Total gas formulas:**
+
+```
+SUCCESS_PATH_TOTAL_GAS =
+    CREATE_FIXED_COST
+  + EXECUTION_GAS
+  + (ACCOUNT_CREATION_COST if account is new else 0)
+  + HASH_GAS(L)
+  + (CODE_DEPOSIT_GAS(L) if not deduped else 0)
+
+FAILURE_PATH_TOTAL_GAS =
+    CREATE_FIXED_COST
+  + EXECUTION_GAS
+```
+
+#### Access-List CodeHash Deduplication
```

Comment:
Wondering if we should add 2930 to the required EIPs field 🤔

Reply:
Hmmm. I thought that this is only done if the EIP required isn't merged. 
Happy to add it. 

Maybe @jochem-brouwer can bring some advice here.